### PR TITLE
feat(datasets) Pre Where Query Processor

### PR DESCRIPTION
**Pull request by fpacifici**
Move the pre-where logic from the view.py file into a Query Processor so that every dataset can provide its own way to provide pre-where conditions.

Besides moving some query processing dataset specific out of view.py, this makes it easy to introduce the new data model in that we would not have logic that specifically modifies the query condition outside of query processors.

test plan:
- added a unit test for this
- the existing pre where test in test_api still works

### CODE REVIEW

**Comment by fpacifici** [comment]
Actually, wait before reviewing this. I forgot a couple of datasets.

**Comment by fpacifici** [comment]
Now it works.

File: snuba/datasets/schemas/__init__.py
Comment (fpacifici) [inline snuba/datasets/schemas/__init__.py]: I moved the pre-where keys as part of the attribute of the data source (whether it is a table, a join or anything else).
This is because pre-where conditions go together with the table we are querying. During query processing we can swap this query, thus the pre where column can change. This way the list is always attached to the table we are querying.

File: snuba/datasets/discover.py
Comment (fpacifici) [inline snuba/datasets/discover.py]: Since the Prewhere candidates are part of the tableSource, I suspect we could have a list of default query processors that are always applied (like this one), no matter on the dataset. For now it is still per dataset.

File: snuba/query/query.py
Comment (fpacifici) [inline snuba/query/query.py]: I want to get back on this and revisit the decision of adding a PREWHERE field to the Snuba query later when I will be working on Clickhouse formatting. Prewhere fields could be processed at that point since PREWHERE is something Clickhouse specific and all the information needed is in the data source, which is accessible from the query. 
Though this means the formatting logic becomes more complex.

File: snuba/datasets/groups.py
Comment (fpacifici) [inline snuba/datasets/groups.py]: Again, this will be better once we have classes for columns and aliases

File: snuba/query/processors/prewhere.py
Comment (lynnagara) [inline snuba/query/processors/prewhere.py]: Don't think this is used

File: snuba/query/processors/prewhere.py
Comment (lynnagara) [inline snuba/query/processors/prewhere.py]: Why do we need to enforce a max of 2 prewhere conditions?

**Review by lynnagara — COMMENTED** [review]
Is it only top level conditions will be moved to prewhere?
For example conditions like `[["event_id", "=", "aaa"], ["event_id", "=", "bbb"]]` i.e. `event_id = aaa OR event_id = bbb` wouldn't be moved to prewhere?

File: snuba/datasets/schemas/__init__.py
Comment (lynnagara) [inline snuba/datasets/schemas/__init__.py]: `candidates`

File: snuba/query/processors/prewhere.py
Comment (fpacifici) [reply snuba/query/processors/prewhere.py]: It is used at line 45.

File: snuba/query/processors/prewhere.py
Comment (fpacifici) [reply snuba/query/processors/prewhere.py]: Don't know. This is a simple cut and paste of the existing prewhere code from views.py. I did not change the loigic.

**Review by fpacifici — COMMENTED** [review]
```
Is it only top level conditions will be moved to prewhere?
For example conditions like [["event_id", "=", "aaa"], ["event_id", "=", "bbb"]] i.e. event_id = aaa OR event_id = bbb wouldn't be moved to prewhere?
```

Yes, it is only the top level conditions. I preserved the logic how it is today.

File: snuba/datasets/schemas/tables.py
Comment (tkaemming) [inline snuba/datasets/schemas/tables.py]:
```suggestion
    def get_prewhere_candidates(self) -> Sequence[str]:
```

File: snuba/datasets/discover.py
Comment (tkaemming) [reply snuba/datasets/discover.py]: I suspect this should eventually be part of a different abstraction than this level of query processing (elaborated in a different comment below) but in general I agree that it should be something that comes along "for free" when using ClickHouse MergeTree-backed storage engines.

File: snuba/datasets/schemas/__init__.py
Comment (tkaemming) [reply snuba/datasets/schemas/__init__.py]: Good observation, this makes a lot of sense.

File: snuba/query/query.py
Comment (fpacifici) [reply snuba/query/query.py]: _Do you have any sense of how much work it would be to move it? _

I think it is complex enough to do it in its own PR, the only reason is that it would change the logic of the prewhere processor itself. Now (and before this PR) the prewhere processing updated the query: it changes the set of conditions. If we do that in the format, we would not update the query object anymore, other wise we go back potentially to that state where formatting the query made changes to it.

So I would keep it separate since this PR does not touch the PREWHERE processing logic.

**Comment by fpacifici** [comment]
Updates:
- naming fixes
- split processors in custom and mandatory. The prewhere processor is a mandatory processor that gets applied to all datasets.

Moving the prewhere processor in will be in a followup PR since it requires the prewhere processing logic to change, this one is isofunctional to reduce the scope of the PR

File: snuba/datasets/dataset.py
Comment (tkaemming) [inline snuba/datasets/dataset.py]: I'm not sure this split is necessary — I think it's fine for datasets to just append or prepend their custom processors to the value returned by `super`.

Ultimately, I don't think this processor this should act on `snuba.query.query.Query` (note: that's a lot of `query`) at all, it should act on `snuba.clickhouse.query.Query` instead (since `{get,set}_prewhere` should be moved there as well — that is what I thought we were talking about [here](https://github.com/getsentry/snuba/pull/574/files#r344992759) in case it wasn't clear), so I'm not sure we're going to get enough long-term value out of this to justify splitting `get_query_processors` and `_get_custom_query_processors`.

**Review by tkaemming — APPROVED** [review]
This looks good to me — I don't think the addition of `_get_custom_query_processors` is necessary (specific reasoning is noted inline) but I don't think it's necessarily a problem, either. Your call if you want to back that out before merging.

File: snuba/datasets/dataset.py
Comment (fpacifici) [reply snuba/datasets/dataset.py]: I saw two issues in the previous version we discussed about (https://github.com/getsentry/snuba/pull/574/files#r344992759):
1) make Prewhere processing automatic for all datasets
2) make Prewhere part of the Clickhouse query format. 

The first is achievable here (thus this method), while the second needs more work. But no strong opinion on this solution since there are only a handful of datasets and making prewhere mandatory is always going to be possible. Will move it back to the get_query_processors method.

**Pull request merged**
